File: src/main/java/com/airsaid/localization/translate/impl/microsoft/MicrosoftTranslator.java
/*
 * Copyright 2021 Airsaid. https://github.com/airsaid
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.airsaid.localization.translate.impl.microsoft;

import com.airsaid.localization.translate.AbstractTranslator;
import com.airsaid.localization.translate.lang.Lang;
import com.airsaid.localization.translate.lang.Languages;
import com.airsaid.localization.translate.util.GsonUtil;
import com.airsaid.localization.translate.util.UrlBuilder;
import com.google.auto.service.AutoService;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.util.io.RequestBuilder;
import icons.PluginIcons;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.ArrayList;
import java.util.List;

/**
 * @author airsaid
 */
@AutoService(AbstractTranslator.class)
public class MicrosoftTranslator extends AbstractTranslator {

  private static final Logger LOG = Logger.getInstance(MicrosoftTranslator.class);

  private static final String KEY = "Microsoft";
  private static final String HOST_URL = "https://api.cognitive.microsofttranslator.com";
  private static final String TRANSLATE_URL = HOST_URL.concat("/translate");
  private static final String APPLY_APP_ID_URL = "https://docs.microsoft.com/azure/cognitive-services/translator/translator-how-to-signup";

  private List<Lang> supportedLanguages;

  @Override
  public @NotNull String getKey() {
    return KEY;
  }

  @Override
  public @NotNull String getName() {
    return "Microsoft";
  }

  @Override
  public @Nullable Icon getIcon() {
    return PluginIcons.MICROSOFT_ICON;
  }

  @Override
  public boolean isNeedAppId() {
    return false;
  }

  @Override
  public @NotNull List<Lang> getSupportedLanguages() {
    if (supportedLanguages == null) {
      supportedLanguages = new ArrayList<>();
      supportedLanguages.add(Languages.AFRIKAANS);
      supportedLanguages.add(Languages.ALBANIAN);
      supportedLanguages.add(Languages.AMHARIC);
      supportedLanguages.add(Languages.ARABIC);
      supportedLanguages.add(Languages.ARMENIAN);
      supportedLanguages.add(Languages.ASSAMESE);
      supportedLanguages.add(Languages.AZERBAIJANI);
      supportedLanguages.add(Languages.BANGLA);
      supportedLanguages.add(Languages.BOSNIAN);
      supportedLanguages.add(Languages.BULGARIAN);
      supportedLanguages.add(Languages.CATALAN);
      supportedLanguages.add(Languages.CHINESE_SIMPLIFIED.setTranslationCode("zh-Hans"));
      supportedLanguages.add(Languages.CHINESE_TRADITIONAL.setTranslationCode("zh-Hant"));
      supportedLanguages.add(Languages.CROATIAN);
      supportedLanguages.add(Languages.CZECH);
      supportedLanguages.add(Languages.DANISH);
      supportedLanguages.add(Languages.DARI);
      supportedLanguages.add(Languages.DUTCH);
      supportedLanguages.add(Languages.ENGLISH);
      supportedLanguages.add(Languages.ESTONIAN);
      supportedLanguages.add(Languages.FIJIAN);
      supportedLanguages.add(Languages.FILIPINO.setTranslationCode("fil"));
      supportedLanguages.add(Languages.FINNISH);
      supportedLanguages.add(Languages.FRENCH);
      supportedLanguages.add(Languages.GERMAN);
      supportedLanguages.add(Languages.GREEK);
      supportedLanguages.add(Languages.GUJARATI);
      supportedLanguages.add(Languages.HAITIAN_CREOLE);
      supportedLanguages.add(Languages.HEBREW.setTranslationCode("he"));
      supportedLanguages.add(Languages.HINDI);
      supportedLanguages.add(Languages.HMONG_DAW);
      supportedLanguages.add(Languages.HUNGARIAN);
      supportedLanguages.add(Languages.ICELANDIC);
      supportedLanguages.add(Languages.INDONESIAN.setTranslationCode("id"));
      supportedLanguages.add(Languages.INUKTITUT);
      supportedLanguages.add(Languages.IRISH);
      supportedLanguages.add(Languages.ITALIAN);
      supportedLanguages.add(Languages.JAPANESE);
      supportedLanguages.add(Languages.KANNADA);
      supportedLanguages.add(Languages.KAZAKH);
      supportedLanguages.add(Languages.KHMER);
      supportedLanguages.add(Languages.KLINGON_LATIN);
      supportedLanguages.add(Languages.KLINGON_PIQAD);
      supportedLanguages.add(Languages.KOREAN);
      supportedLanguages.add(Languages.KURDISH);
      supportedLanguages.add(Languages.LAO);
      supportedLanguages.add(Languages.LATVIAN);
      supportedLanguages.add(Languages.LITHUANIAN);
      supportedLanguages.add(Languages.MALAGASY);
      supportedLanguages.add(Languages.MALAY);
      supportedLanguages.add(Languages.MALAYALAM);
      supportedLanguages.add(Languages.MALTESE);
      supportedLanguages.add(Languages.MAORI);
      supportedLanguages.add(Languages.MARATHI);
      supportedLanguages.add(Languages.BURMESE);
      supportedLanguages.add(Languages.NEPALI);
      supportedLanguages.add(Languages.NORWEGIAN.setTranslationCode("nb"));
      supportedLanguages.add(Languages.ODIA);
      supportedLanguages.add(Languages.PASHTO);
      supportedLanguages.add(Languages.PERSIAN);
      supportedLanguages.add(Languages.PORTUGUESE);
      supportedLanguages.add(Languages.PUNJABI);
      supportedLanguages.add(Languages.QUERETARO_OTOMI);
      supportedLanguages.add(Languages.ROMANIAN);
      supportedLanguages.add(Languages.RUSSIAN);
      supportedLanguages.add(Languages.SAMOAN);
      supportedLanguages.add(Languages.SERBIAN);
      supportedLanguages.add(Languages.SLOVAK);
      supportedLanguages.add(Languages.SLOVENIAN);
      supportedLanguages.add(Languages.SPANISH);
      supportedLanguages.add(Languages.SWAHILI);
      supportedLanguages.add(Languages.SWEDISH);
      supportedLanguages.add(Languages.TAHITIAN);
      supportedLanguages.add(Languages.TAMIL);
      supportedLanguages.add(Languages.TELUGU);
      supportedLanguages.add(Languages.THAI);
      supportedLanguages.add(Languages.TIGRINYA);
      supportedLanguages.add(Languages.TONGAN);
      supportedLanguages.add(Languages.TURKISH);
      supportedLanguages.add(Languages.UKRAINIAN);
      supportedLanguages.add(Languages.URDU);
      supportedLanguages.add(Languages.VIETNAMESE);
      supportedLanguages.add(Languages.WELSH);
      supportedLanguages.add(Languages.YUCATEC_MAYA);
    }
    return supportedLanguages;
  }

  @Override
  public String getAppKeyDisplay() {
    return "KEY";
  }

  @Override
  public @Nullable String getApplyAppIdUrl() {
    return APPLY_APP_ID_URL;
  }

  @Override
  public @NotNull String getRequestUrl(@NotNull Lang fromLang, @NotNull Lang toLang, @NotNull String text) {
    return new UrlBuilder(TRANSLATE_URL)
        .addQueryParameter("api-version", "3.0")
        .addQueryParameter("to", toLang.getTranslationCode())
        .build();
  }

  @Override
  @NotNull
  public String getRequestBody(@NotNull Lang fromLang, @NotNull Lang toLang, @NotNull String text) {
    return "[{\"Text\": \"" + text + "\"}]";
  }

  @Override
  public void configureRequestBuilder(@NotNull RequestBuilder requestBuilder) {
    requestBuilder.tuner(connection -> {
      connection.setRequestProperty("Ocp-Apim-Subscription-Key", getAppKey());
      connection.setRequestProperty("Content-type", "application/json");
    });
  }

  @Override
  public @NotNull String parsingResult(@NotNull Lang fromLang, @NotNull Lang toLang, @NotNull String text, @NotNull String resultText) {
    LOG.info("parsingResult: " + resultText);
    return GsonUtil.getInstance().getGson().fromJson(resultText, MicrosoftTranslationResult[].class)[0].getTranslationResult();
  }

}

File: src/main/kotlin/com/airsaid/localization/translate/impl/microsoft/MicrosoftTranslator.kt
/*
 * Copyright 2021 Airsaid. https://github.com/airsaid
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.airsaid.localization.translate.impl.microsoft

import com.airsaid.localization.translate.AbstractTranslator
import com.airsaid.localization.translate.TranslatorCredentialDescriptor
import com.airsaid.localization.translate.lang.Lang
import com.airsaid.localization.translate.lang.Languages
import com.airsaid.localization.translate.lang.toLang
import com.airsaid.localization.translate.util.GsonUtil
import com.airsaid.localization.translate.util.UrlBuilder
import com.google.auto.service.AutoService
import com.intellij.openapi.diagnostic.Logger
import com.intellij.util.io.RequestBuilder
import icons.PluginIcons

/**
 * @author airsaid
 */
@AutoService(AbstractTranslator::class)
class MicrosoftTranslator : AbstractTranslator() {

  companion object {
    private val LOG = Logger.getInstance(MicrosoftTranslator::class.java)
    private const val KEY = "Microsoft"
    private const val DEFAULT_HOST_URL = "https://api.cognitive.microsofttranslator.com"

    @Volatile
    internal var hostOverride: String? = null

    private val HOST_URL: String
      get() = hostOverride ?: DEFAULT_HOST_URL

    private val TRANSLATE_URL: String
      get() = "$HOST_URL/translate"
  }

  override val key = KEY

  override val icon = PluginIcons.MICROSOFT_ICON

  override val credentialDefinitions: List<TranslatorCredentialDescriptor> = emptyList()

  override val credentialHelpUrl: String? = null

  override val supportedLanguages: List<Lang> by lazy {
    buildList {
      add(Languages.AFRIKAANS.toLang())
      add(Languages.ALBANIAN.toLang())
      add(Languages.AMHARIC.toLang())
      add(Languages.ARABIC.toLang())
      add(Languages.ARMENIAN.toLang())
      add(Languages.ASSAMESE.toLang())
      add(Languages.AZERBAIJANI.toLang())
      add(Languages.BANGLA.toLang())
      add(Languages.BOSNIAN.toLang())
      add(Languages.BULGARIAN.toLang())
      add(Languages.CATALAN.toLang())
      add(Languages.CHINESE_SIMPLIFIED.toLang().setTranslationCode("zh-Hans"))
      add(Languages.CHINESE_TRADITIONAL.toLang().setTranslationCode("zh-Hant"))
      add(Languages.CROATIAN.toLang())
      add(Languages.CZECH.toLang())
      add(Languages.DANISH.toLang())
      add(Languages.DARI.toLang())
      add(Languages.DUTCH.toLang())
      add(Languages.ENGLISH.toLang())
      add(Languages.ESTONIAN.toLang())
      add(Languages.FIJIAN.toLang())
      add(Languages.FILIPINO.toLang().setTranslationCode("fil"))
      add(Languages.FINNISH.toLang())
      add(Languages.FRENCH.toLang())
      add(Languages.GERMAN.toLang())
      add(Languages.GREEK.toLang())
      add(Languages.GUJARATI.toLang())
      add(Languages.HAITIAN_CREOLE.toLang())
      add(Languages.HEBREW.toLang().setTranslationCode("he"))
      add(Languages.HINDI.toLang())
      add(Languages.HMONG_DAW.toLang())
      add(Languages.HUNGARIAN.toLang())
      add(Languages.ICELANDIC.toLang())
      add(Languages.INDONESIAN.toLang().setTranslationCode("id"))
      add(Languages.INUKTITUT.toLang())
      add(Languages.IRISH.toLang())
      add(Languages.ITALIAN.toLang())
      add(Languages.JAPANESE.toLang())
      add(Languages.KANNADA.toLang())
      add(Languages.KAZAKH.toLang())
      add(Languages.KHMER.toLang())
      add(Languages.KOREAN.toLang())
      add(Languages.KURDISH.toLang())
      add(Languages.LAO.toLang())
      add(Languages.LATVIAN.toLang())
      add(Languages.LITHUANIAN.toLang())
      add(Languages.MALAGASY.toLang())
      add(Languages.MALAY.toLang())
      add(Languages.MALAYALAM.toLang())
      add(Languages.MALTESE.toLang())
      add(Languages.MAORI.toLang())
      add(Languages.MARATHI.toLang())
      add(Languages.BURMESE.toLang())
      add(Languages.NEPALI.toLang())
      add(Languages.NORWEGIAN.toLang().setTranslationCode("nb"))
      add(Languages.ODIA.toLang())
      add(Languages.PASHTO.toLang())
      add(Languages.PERSIAN.toLang())
      add(Languages.PORTUGUESE.toLang())
      add(Languages.PUNJABI.toLang())
      add(Languages.QUERETARO_OTOMI.toLang())
      add(Languages.ROMANIAN.toLang())
      add(Languages.RUSSIAN.toLang())
      add(Languages.SAMOAN.toLang())
      add(Languages.SERBIAN.toLang())
      add(Languages.SLOVAK.toLang())
      add(Languages.SLOVENIAN.toLang())
      add(Languages.SPANISH.toLang())
      add(Languages.SWAHILI.toLang())
      add(Languages.SWEDISH.toLang())
      add(Languages.TAHITIAN.toLang())
      add(Languages.TAMIL.toLang())
      add(Languages.TELUGU.toLang())
      add(Languages.THAI.toLang())
      add(Languages.TIGRINYA.toLang())
      add(Languages.TONGAN.toLang())
      add(Languages.TURKISH.toLang())
      add(Languages.UKRAINIAN.toLang())
      add(Languages.URDU.toLang())
      add(Languages.VIETNAMESE.toLang())
      add(Languages.WELSH.toLang())
      add(Languages.YUCATEC_MAYA.toLang())
    }
  }

  override val requestContentType = "application/json"

  override fun getRequestUrl(fromLang: Lang, toLang: Lang, text: String): String {
    return UrlBuilder(TRANSLATE_URL)
      .addQueryParameter("api-version", "3.0")
      .addQueryParameter("to", toLang.translationCode)
      .build()
  }

  override fun getRequestBody(fromLang: Lang, toLang: Lang, text: String): String {
    return "[{\"Text\": \"$text\"}]"
  }

  override fun configureRequestBuilder(requestBuilder: RequestBuilder) {
    requestBuilder.tuner { connection ->
      val token = MicrosoftEdgeAuthService.getInstance().getAccessToken()
      connection.setRequestProperty("Authorization", "Bearer $token")
      connection.setRequestProperty("Content-type", "application/json")
    }
  }

  override fun parsingResult(fromLang: Lang, toLang: Lang, text: String, resultText: String): String {
    LOG.info("parsingResult: $resultText")
    return GsonUtil.getInstance().gson.fromJson(
      resultText,
      Array<MicrosoftTranslationResult>::class.java
    )[0].translationResult
  }
}
